### ISSUE TITLE
Playing Local File instead of Web File?

### ISSUE THREAD

**Issue by darethehair** [erkstruwe/chromecast-cli#15]
Is there a way to play a local file from this utility -- rather than from a URL?

```
darren@dell3010:~$ chromecast --host 192.168.4.35 play nosuchfile.mp3
Playing nosuchfile.mp3
Status IDLE
```

The file 'nosuchfile.mp3' doesn't actually exist, yet the messages make it appear as though it has found the file and successfully played it.

**Comment by erkstruwe** [comment]
That's not possible because the file in question must be accessible on a network. It is streamed rather than uploaded all at once.

**Comment by darethehair** [comment]
Understood :)  

Out of curiosity, I used the "Webserver for Chrome" (https://chromewebstore.google.com/detail/web-server-for-chrome/ofhbbkphhbklhfoeikjpcbhemlocgigb?pli=1) to see if I could make that work for me, which creates a URL as follows:

`http://127.0.0.1:8887`

I used this to create a reference to the sample MP3 file like this, but it did not work:

`http://127.0.0.1:8887/test.mp3`

Why am I doing this?  Since our local non-profit uses ChromeOS (Flex), with a bunch of Chromecasts, and we need/use this webserver to play some kiosks successfuly.  Within the local Linux virtual system (Crostini), I am able to use your utility to play MP3 files (though not locally).  I want to continue to host the MP3 files on our machines directly, and I hoped that this 'webserver' would allow me to honor the requirements of your utility to do so :)

**Comment by erkstruwe** [comment]
The IP address `127.0.0.1` refers to the machine itself and is not accessible from any other machine. If you can access the file at the URL http://127.0.0.1:8887/test.mp3, all you have to do is find out that machine's local IP address on the network. The creator of "Webserver for Chrome" shows it in this video: https://www.youtube.com/watch?v=AK6swHiPtew. Watch out for him mentioning "Accessible by other computers".

So, basically, if you have a URL like http://192.168.1.124:8887/test.mp3 (one that is accessible from the network in which the Chromecasts reside), it should work.

BTW: The files having to be accessible on a network is not a requirement of `chromecast-cli` but of the entire "Cast" functionality in general.

**Comment by darethehair** [comment]
Just an FYI: There must be something a bit strange about using webservers *within* ChromeOS Crostini, since using the non-local-host technique with "Webserver for Chrome" as described above is not sufficient.  Instead, I need to use something like 'ngrok' to create an alternate external IP address for the internal web server before utilities like your own become successful.

e.g. This works:

chromecast --host 192.168.4.35 play https://abc123whatever.ngrok-free.app/test.mp3

I also found an alternate webserver that (at least in this case) works even better for serving my MP3 file than "Webserver for Chrome".

i.e. python3 -m http.server

For some reason, this one seems to work just fine with a '127.0.0.0' localhost IP.  However, it still requires ngrok in order to play it within Crostini.

**Comment by erkstruwe** [comment]
Hi!

It seems to me that by doing this you make the files available on the internet, which is not always feasible.

I just tested the Chrome app you mentioned in my local setup at home and everything works great. I checked "Accessible on local network", restarted the web server and chose the URL from the list of "Web Server URL(s)" that is on the same network as the Chromecast. Then, by entering this (on another machine), I could play a demo file on my Chromecast streamed from my Chromebook:
`chromecast --host 192.168.1.20 play http://192.168.1.19:8887/file_example_MP3_700KB.mp3`

`192.168.1.20` is the local IP address of my Chromecast. `192.168.1.19` is the local IP address of my Chromebook, `8887` is the port set in "Web Server for Chrom". Since Chrome Apps won't be available from 2025, I'd suggest opting for a more sustainable solution. Any web server application should suffice.

**Comment by darethehair** [comment]
Yes, I confirmed that the "Web Server for Chrome" technique does work in my normal Linux environment, but it does not work within the ChromeOS Crostini Linux VM that I need it for i.e. the 'internal network' there is (for me) '100.115.92.xxx' rather than the normal '192.168.4.xxx' that my normal systems use.  This, then, is not the fault of your utility :)

```
darren@penguin:~$ chromecast --host 192.168.4.35 play http://100.115.92.129:8887/test.mp3
Playing http://100.115.92.129:8887/test.mp3
Status IDLE
```
(no error messages, but no sound)

I've also tried a variation of this by using the 'port forwarding' feature of Crostini (same port '8887' as above), in order to turn the IP address into something that looks normal:

```
darren@penguin:~$ chromecast --host 192.168.4.35 play http://192.168.5.45:8887/test.mp3
Playing http://192.168.5.45:8887/test.mp3
Status IDLE
```
Same result -- no sound and no errors -- even though that webserver address contains the file and plays normally through the browser.

This goes beyond my own technical skills, but it seems to me that it *should* work but just does not.  Thanks for your help in any case!  

Yes, I am concerned that Chrome Apps are going away -- though I still hope that ChromeOS itself will avoid this fate. I have made all possible efforts to support my local thrift shop using *only* ChromeOS -- in order to make it the easiest to maintain when I am no longer able to volunteer there, and for the most part I have been successful -- but lack of a web server could be tricky, and I might need to install a 'real' webserver to see if that works.

UPDATE!  Got it to work by using a particular configuration of python3 http server!

```
darren@penguin:~$ python3 -m http.server 8000
Serving HTTP on 0.0.0.0 port 8000 (http://0.0.0.0:8000/) ...
192.168.4.35 - - [26/Jan/2024 10:17:54] "GET /test.mp3 HTTP/1.1" 200 -
192.168.4.76 - - [26/Jan/2024 10:18:21] "GET / HTTP/1.1" 200 -
192.168.4.76 - - [26/Jan/2024 10:18:21] code 404, message File not found
192.168.4.76 - - [26/Jan/2024 10:18:21] "GET /favicon.ico HTTP/1.1" 404 -
```

When I used the '0.0.0.0' configuration, your Chromecast utiility worked:

```
darren@penguin:~$ chromecast --host 192.168.4.35 play http://192.168.5.45:8000/test.mp3
Playing http://192.168.5.45:8000/test.mp3
Status IDLE
```

I got this 'hint' about using '0.0.0.0' from here, but I am not sure exactly what it means:

[https://www.reddit.com/r/Crostini/comments/hl96yg/using_crostini_ssh_port_forwarding_to_forward/](url)

**Comment by erkstruwe** [comment]
Ah, I see. So if you run the web server in Crostini, Crostini's IP address is only accessible by the Chromebook itself. However, it seems to me that [Chrome OS's port forwarding](https://chromeos.dev/en/web-environment/port-forwarding) is meant to solve exactly that. You must forward an arbitrary Chrome OS port (e. g. `8080`) to the port of the web server you launched in Crostini (e. g. `80` or so, depending on the web server and its settings used). Then, use Chrome OS's IP address along with Chrome OS's port: `http://192.168.5.45:8080/test.mp3`

**Comment by darethehair** [comment]
> Ah, I see. So if you run the web server in Crostini, Crostini's IP address is only accessible by the Chromebook itself. However, it seems to me that [Chrome OS's port forwarding](https://chromeos.dev/en/web-environment/port-forwarding) is meant to solve exactly that. You must forward an arbitrary Chrome OS port (e. g. `8080`) to the port of the web server you launched in Crostini (e. g. `80` or so, depending on the web server and its settings used). Then, use Chrome OS's IP address along with Chrome OS's port: `http://192.168.5.45:8080/test.mp3`

Yes, as I indicated, the 'port forwarding' feature works -- but (apparently) not with 'Webserver for Chrome' -- but it does seem to work with the python3 built-in webserver.  Lots for me to try to understand here, but this is progress! :)
